app-creator/src/redux/reducer.ts

```diff
@@ -61,23 +68,19 @@ export const reducer: Reducer<AppCreatorStore, AppCreatorAction | AnyAction> = (
     case SET_DRAGGING_WIDGET:
       return {
         ...state,
-        element: action.payload.widget,
-        eventType: EventType.none,
+        ...action.payload,
       };
     case SET_EDITING_WIDGET:
       return {
         ...state,
-        element: action.payload.widget,
-        eventType: EventType.editing,
+        ...action.payload,
         selectedTab:
-          action.payload.widget == null
-            ? state.selectedTab
-            : action.payload.index,
+          action.payload.element == null ? state.selectedTab : Tab.attributes,
```

Comment:
Using payload element to determine the selected tab is a bit indirect and could be confusing to future reviewers. Could you find another way of determining whether we should open the attribute tab, or add a comment to explain it?

Reply:
+1, it may be worth adding a comment below every case statement about the purpose of the reducer.

Reply:
I'm thinking of doing something like this 
```Typescript
export const setEditingWidget = (
  widget: Element | null
): SetEditingWidgetAction => {
  return {
    type: SET_EDITING_WIDGET,
    payload: {
      element: widget,
      // If widget is null, then we want to clear the editing state.
      eventType: widget == null ? EventType.none : EventType.editing,
      // Open attributes tab if we are editing an element (ie. Not clearing state).
      openAttributesTab: widget != null,
    },
  };
};
```

And then have this in the reducer

```Typescript
case SET_EDITING_WIDGET:
      return {
        ...state,
        ...action.payload,
        selectedTab: action.payload.openAttributesTab
          ? Tab.attributes
          : state.selectedTab,
      };
```

Alternatively, we can separate the editing action into two different actions
1. `SetEditingWidget`
2. `ClearEditingWidget`

Reply:
That sounds good! I would also add an explanation about when/why a widget could be null, next to the comment `// If widget is null, then we want to clear the editing state.`